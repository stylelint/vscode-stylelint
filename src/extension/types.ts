import type TypedEventEmitter from 'typed-emitter';
import type { DidRegisterDocumentFormattingEditProviderNotificationParams } from '../server/index';

/**
 * VS Code extension event names.
 */
export type ExtensionEvents = {
	DidRegisterDocumentFormattingEditProvider: (
		params: DidRegisterDocumentFormattingEditProviderNotificationParams,
	) => void;
	DidResetConfiguration: () => void;
};

/**
 * VS Code extension public API.
 */
export type PublicApi = TypedEventEmitter<ExtensionEvents> & {
	codeActionReady: boolean;
};

/**
 * Extension API event names.
 */
export enum ApiEvent {
	DidRegisterDocumentFormattingEditProvider = 'DidRegisterDocumentFormattingEditProvider',
	DidResetConfiguration = 'DidResetConfiguration',
}
