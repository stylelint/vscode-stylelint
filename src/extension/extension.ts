import { EventEmitter } from 'node:events';
import path from 'node:path';

import { LanguageClient, SettingMonitor, ExecuteCommandRequest } from 'vscode-languageclient/node';
import { workspace, commands, window, type ExtensionContext } from 'vscode';
import { ApiEvent, PublicApi } from './types';
import {
	CommandId,
	DidRegisterDocumentFormattingEditProviderNotificationParams,
	Notification,
} from '../server/index';

let client: LanguageClient;

/**
 * Activates the extension.
 */
export async function activate({ subscriptions }: ExtensionContext): Promise<PublicApi> {
	const serverPath = path.join(__dirname, 'start-server.js');

	const api = Object.assign(new EventEmitter(), {
		codeActionReady: false,
	}) as PublicApi;

	client = new LanguageClient(
		'Stylelint',
		{
			run: {
				module: serverPath,
			},
			debug: {
				module: serverPath,
				options: {
					execArgv: ['--nolazy', '--inspect=6004'],
				},
			},
		},
		{
			documentSelector: [{ scheme: 'file' }, { scheme: 'untitled' }],
			diagnosticCollectionName: 'Stylelint',
			synchronize: {
				fileEvents: [
					workspace.createFileSystemWatcher('**/.stylelintrc{,.js,.json,.yaml,.yml}'),
					workspace.createFileSystemWatcher('**/{stylelint.config.js,.stylelintignore}'),
				],
			},
		},
	);

	const errorHandler = async (error: unknown): Promise<void> => {
		await window.showErrorMessage(
			`Stylelint: ${error instanceof Error ? error.message : String(error)}`,
		);
	};

	const notificationHandlers = (): void => {
		client.onNotification(Notification.DidRegisterCodeActionRequestHandler, () => {
			api.codeActionReady = true;
		});
		client.onNotification(
			Notification.DidRegisterDocumentFormattingEditProvider,
			(params: DidRegisterDocumentFormattingEditProviderNotificationParams) => {
				api.emit(ApiEvent.DidRegisterDocumentFormattingEditProvider, params);
			},
		);
		client.onNotification(Notification.DidResetConfiguration, () => {
			api.emit(ApiEvent.DidResetConfiguration);
		});
	};

	try {
		await client.start();
		notificationHandlers();
	} catch (err) {
		await errorHandler(err);
	}

	subscriptions.push(
		// cspell:disable-next-line
		commands.registerCommand('stylelint.executeAutofix', async () => {
			const textEditor = window.activeTextEditor;

			if (!textEditor) {
				return;
			}

			const textDocument = {
				uri: textEditor.document.uri.toString(),
				version: textEditor.document.version,
			};
			const params = {
				command: CommandId.ApplyAutoFix,
				arguments: [textDocument],
			};

			try {
				await client.sendRequest(ExecuteCommandRequest.type, params);
			} catch {
				await window.showErrorMessage(
					'Failed to apply Stylelint fixes to the document. Please consider opening an issue with steps to reproduce.',
				);
			}
		}),
	);

	subscriptions.push(
		commands.registerCommand('stylelint.restart', async () => {
			await client.stop();

			try {
				await client.start();
				notificationHandlers();
			} catch (error: unknown) {
				await errorHandler(error);
			}
		}),
	);

	subscriptions.push(new SettingMonitor(client, 'stylelint.enable').start());

	return Promise.resolve(api);
}

/**
 * @returns A promise that resolves when the client has been deactivated.
 */
export async function deactivate(): Promise<void> {
	if (client) {
		try {
			return client.stop();
		} catch (err) {
			const msg = err && (err as Error) ? (err as Error).message : 'unknown';

			await window.showErrorMessage(`error stopping stylelint language server: ${msg}`);
			throw err;
		}
	}
}
