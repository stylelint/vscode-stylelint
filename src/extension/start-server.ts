import path from 'path';
import process from 'process';
import { createConnection, ProposedFeatures } from 'vscode-languageserver/node';
import { StylelintLanguageServer, modules, createLogger } from '../server/index';

const connection = createConnection(ProposedFeatures.all);

const { NODE_ENV } = process.env;

const logger =
	NODE_ENV === 'development'
		? createLogger(connection, 'debug', path.join(__dirname, '../stylelint-language-server.log'))
		: createLogger(connection, 'info');

const server = new StylelintLanguageServer({
	connection,
	logger,
	modules: Object.values(modules),
});

server.start();
