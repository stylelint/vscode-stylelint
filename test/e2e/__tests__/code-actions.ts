/* eslint-disable jest/no-standalone-expect */
import fs from 'fs/promises';
import path from 'path';
import process from 'process';

import pWaitFor from 'p-wait-for';
import {
	commands,
	extensions,
	workspace,
	Selection,
	Range,
	Position,
	CodeAction,
	TextEditor,
} from 'vscode';
import { ApiEvent, PublicApi } from '../../../src/extension/index';

const getCodeActions = async (editor: TextEditor): Promise<CodeAction[]> =>
	(await commands.executeCommand(
		'vscode.executeCodeActionProvider',
		editor.document.uri,
		new Range(editor.selection.start, editor.selection.end),
	)) ?? [];

const serializeCodeActions = (actions: CodeAction[]) =>
	actions.map((action) => ({
		...action,
		...(action.edit ? { edit: action.edit?.entries()?.map(([, edits]) => ['<uri>', edits]) } : {}),
	}));

const cssPath = path.resolve(workspaceDir, 'code-actions/test.css');
const jsPath = path.resolve(workspaceDir, 'code-actions/test.js');
const settingsPath = path.resolve(workspaceDir, 'code-actions/.vscode/settings.json');

// TODO: Investigate why editing tests intermittently fail on CI
const localIt = process.env.CI ? it.skip : it;

describe('Code actions', () => {
	beforeAll(async () => {
		const api = (await extensions.getExtension('stylelint.vscode-stylelint')?.exports) as PublicApi;

		await pWaitFor(() => api.codeActionReady, { timeout: 5000 });
	});

	let savedFiles: Map<string, string>;

	beforeEach(async () => {
		savedFiles = new Map([
			[cssPath, await fs.readFile(cssPath, 'utf8')],
			[settingsPath, await fs.readFile(settingsPath, 'utf8')],
		]);
	});

	afterEach(async () => {
		for (const [filePath, content] of savedFiles.entries()) {
			await fs.writeFile(filePath, content);
		}
	});

	it('should provide code actions for problems', async () => {
		const editor = await openDocument(cssPath);

		await waitForDiagnostics(editor);

		editor.selection = new Selection(new Position(1, 11), new Position(1, 11));

		const actions = await getCodeActions(editor);

		expect(serializeCodeActions(actions)).toMatchSnapshot();
	});

	it('should not provide disable code actions for disable reports', async () => {
		const editor = await openDocument(cssPath);

		await waitForDiagnostics(editor);

		editor.selection = new Selection(new Position(2, 2), new Position(2, 2));

		const actions = await getCodeActions(editor);

		expect(actions).toHaveLength(0);
	});

	test('should run auto-fix action on save', async () => {
		const editor = await openDocument(cssPath);

		await waitForDiagnostics(editor);

		await getCodeActions(editor);

		// API won't save unless we dirty the document, unlike saving via the UI
		await editor.edit((editBuilder) => {
			editBuilder.insert(new Position(3, 0), ' ');
		});

		await editor.document.save();

		expect(editor.document.getText()).toMatchSnapshot();
	});

	localIt('should disable rules for an entire file', async () => {
		const editor = await openDocument(cssPath);

		await waitForDiagnostics(editor);

		editor.selection = new Selection(new Position(1, 11), new Position(1, 11));

		const actions = await getCodeActions(editor);

		const fileAction = actions.find((action) =>
			action.title.match(/^Disable .+ for the entire file$/),
		);

		expect(fileAction?.edit).toBeDefined();

		await workspace.applyEdit(fileAction!.edit!);

		expect(editor.document.getText()).toMatchSnapshot();
	});

	localIt('should disable rules for an entire file with a shebang', async () => {
		const editor = await openDocument(jsPath);

		await waitForDiagnostics(editor);

		editor.selection = new Selection(new Position(6, 9), new Position(6, 9));

		const actions = await getCodeActions(editor);

		const fileAction = actions.find((action) =>
			action.title.match(/^Disable .+ for the entire file$/),
		);

		expect(fileAction?.edit).toBeDefined();

		await workspace.applyEdit(fileAction!.edit!);

		expect(editor.document.getText()).toMatchSnapshot();
	});

	localIt(
		'should disable rules for a specific line with a comment on the previous line',
		async () => {
			const editor = await openDocument(cssPath);

			editor.selection = new Selection(new Position(1, 11), new Position(1, 11));

			await waitForDiagnostics(editor);

			const actions = await getCodeActions(editor);
			const lineAction = actions.find((action) => action.title.match(/^Disable .+ for this line$/));

			expect(lineAction?.edit).toBeDefined();

			await workspace.applyEdit(lineAction!.edit!);

			expect(editor.document.getText()).toMatchSnapshot();
		},
	);

	localIt('should disable rules for a specific line with a comment on the same line', async () => {
		const settingsEditor = await openDocument(settingsPath);

		await settingsEditor.edit((edit) =>
			edit.insert(
				new Position(10, 2),
				',\n\t"stylelint.codeAction.disableRuleComment": { "location": "sameLine" }',
			),
		);

		const resetPromise = waitForApiEvent(ApiEvent.DidResetConfiguration);

		await settingsEditor.document.save();
		await resetPromise;
		//await openDocument(cssPath);
		//await commands.executeCommand('workbench.action.closeActiveEditor');

		const editor = await openDocument(cssPath);

		await waitForDiagnostics(editor);

		editor.selection = new Selection(new Position(1, 11), new Position(1, 11));

		const actions = await getCodeActions(editor);
		const lineAction = actions.find((action) => action.title.match(/^Disable .+ for this line$/));

		expect(lineAction?.edit).toBeDefined();

		await workspace.applyEdit(lineAction!.edit!);

		expect(editor.document.getText()).toMatchSnapshot();
	});
});
