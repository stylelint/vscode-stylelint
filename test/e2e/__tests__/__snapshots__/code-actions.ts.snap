// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Code actions should disable rules for a specific line with a comment on the previous line 1`] = `
"a {
  /* stylelint-disable-next-line value-keyword-case */
  display: BLOCK;
  /* stylelint-disable-next-line comment-no-empty */
  color: #00;
}
"
`;

exports[`Code actions should disable rules for a specific line with a comment on the same line 1`] = `
"a {
  display: BLOCK; /* stylelint-disable-line value-keyword-case */
  /* stylelint-disable-next-line comment-no-empty */
  color: #00;
}
"
`;

exports[`Code actions should disable rules for an entire file 1`] = `
"/* stylelint-disable value-keyword-case */
a {
  display: BLOCK;
  /* stylelint-disable-next-line comment-no-empty */
  color: #00;
}
"
`;

exports[`Code actions should disable rules for an entire file with a shebang 1`] = `
"#!/usr/bin/env node
/* stylelint-disable color-no-invalid-hex */
/* eslint-disable node/shebang */
'use strict';

const css = css\`
	.foo {
		color: #00;
	}
\`;
"
`;

exports[`Code actions should provide code actions for problems 1`] = `
[
  {
    "edit": [
      [
        "<uri>",
        [
          {
            "newEol": undefined,
            "newText": "  /* stylelint-disable-next-line value-keyword-case */
",
            "range": [
              {
                "character": 0,
                "line": 1,
              },
              {
                "character": 0,
                "line": 1,
              },
            ],
          },
        ],
      ],
    ],
    "isPreferred": undefined,
    "kind": _ {
      "value": "quickfix",
    },
    "title": "Disable value-keyword-case for this line",
  },
  {
    "edit": [
      [
        "<uri>",
        [
          {
            "newEol": undefined,
            "newText": "/* stylelint-disable value-keyword-case */
",
            "range": [
              {
                "character": 0,
                "line": 0,
              },
              {
                "character": 0,
                "line": 0,
              },
            ],
          },
        ],
      ],
    ],
    "isPreferred": undefined,
    "kind": _ {
      "value": "quickfix",
    },
    "title": "Disable value-keyword-case for the entire file",
  },
  {
    "command": {
      "arguments": [
        {
          "uri": "https://stylelint.io/user-guide/rules/value-keyword-case",
        },
      ],
      "command": "stylelint.openRuleDoc",
      "title": "Open documentation for value-keyword-case",
    },
    "isPreferred": undefined,
    "kind": _ {
      "value": "quickfix",
    },
    "title": "Show documentation for value-keyword-case",
  },
]
`;

exports[`Code actions should run auto-fix action on save 1`] = `
"a {
  display: block;
  /* stylelint-disable-next-line comment-no-empty */
   color: #00;
}
"
`;
